### PR TITLE
[Internal] Thin Client Integration: Removes ThinClientStoreModel and uses GatewayStoreModel for better code modularity.

### LINKED ISSUE #5311: [Thin Client Integration] Redundant Store Model: Remove the `ThinClientStoreModel` and use `GatewayStoreModel` for Better Code Modularity
### Acceptance Criteria:

As of today, it is observed that most of the working principle on the `ThinClientStoreModel` can be achieved by the `GatewayStoreModel`. Therefore, it would make sense to remove the dependency on the `ThinClientStoreModel` and use the `GatewayStoreModel` instead.

Please make necessary code changes and write unit tests to cover the refactoring.

### PR DESCRIPTION
**Pull request by aavasthy**
# Pull Request Template

## Description

It is observed that most of the working principle on the ThinClientStoreModel can be achieved by the GatewayStoreModel. Therefore, removing ThinClientStoreModel and accommodating the logic to navigate between thinclient mode and gateway mode in GatewayStoreModel. Added tests to verify correct working of all three connection modes: thinclient, gateway and direct.

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

## Closing issues

To automatically close an issue: closes #5311

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/DocumentClient.cs [hunk 1]
```diff
@@ -1074,68 +1074,50 @@ private async Task<bool> GetInitializationTaskAsync(IStoreClientFactory storeCli
                         this.isThinClientEnabled)
                     : GlobalPartitionEndpointManagerNoOp.Instance;
 
-            this.retryPolicy = new RetryPolicy(
-                globalEndpointManager: this.GlobalEndpointManager,
-                connectionPolicy: this.ConnectionPolicy,
+            this.retryPolicy = new RetryPolicy(
```
Comment (kundadebdatta) [inline Microsoft.Azure.Cosmos/src/DocumentClient.cs]: Are these changes related ? If not - can we please clean them up ?

File: Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs [hunk 2]
```diff
@@ -30,93 +30,146 @@ internal class GatewayStoreModel : IStoreModelExtension, IDisposable
...
+        public GatewayStoreModel(
+             GlobalEndpointManager endpointManager,
+             ISessionContainer sessionContainer,
+             ConsistencyLevel defaultConsistencyLevel,
+             DocumentClientEventSource eventSource,
+             JsonSerializerSettings serializerSettings,
+             CosmosHttpClient httpClient,
+             GlobalPartitionEndpointManager globalPartitionEndpointManager,
+             bool isPartitionLevelFailoverEnabled = false,
+             bool enableThinClientMode = false,
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs]: Why default it? Callers can define it always right?

File: Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs [hunk 3]
```diff
@@ -30,93 +30,146 @@ internal class GatewayStoreModel : IStoreModelExtension, IDisposable
...
+                this.eventSource,
+                serializerSettings,
                 isPartitionLevelFailoverEnabled);
-
-            this.globalPartitionEndpointManager.SetBackgroundConnectionPeriodicRefreshTask(
+
+            this.enableThinClientMode = enableThinClientMode;
+            this.userAgentContainer = userAgentContainer;
+
+            if (enableThinClientMode && userAgentContainer != null)
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs]: Can you please expand on why UserAgent is needed in conditions?

File: Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs [hunk 4]
```diff
@@ -30,93 +30,146 @@ internal class GatewayStoreModel : IStoreModelExtension, IDisposable
...
+            {
+                request.RequestContext.RegionName = regionName;
+            }
+
+            // This is applicable for both per partition automatic failover and per partition circuit breaker.
+            if (this.isPartitionLevelFailoverEnabled
+                && !ReplicatedResourceClient.IsMasterResource(request.ResourceType)
+                && request.ResourceType.IsPartitioned())
+            {
+                (bool isSuccess, PartitionKeyRange partitionKeyRange) = await TryResolvePartitionKeyRangeAsync(
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs]: Try-catch is now not wrapped around it right?

File: Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs [hunk 5]
```diff
@@ -30,93 +30,146 @@ internal class GatewayStoreModel : IStoreModelExtension, IDisposable
...
+                    sessionContainer: this.sessionContainer,
+                    partitionKeyRangeCache: this.partitionKeyRangeCache,
+                    clientCollectionCache: this.clientCollectionCache,
+                    refreshCache: false);
+
+                request.RequestContext.ResolvedPartitionKeyRange = partitionKeyRange;
+                this.globalPartitionEndpointManager.TryAddPartitionLevelLocationOverride(request);
+            }
+
+            Uri thinClientEndpoint = this.endpointManager.ResolveThinClientEndpoint(request);
```
Comment (kundadebdatta) [inline Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs]: Resolve `thinClientEndpoint` under `if (canUseThinClient)` check within the `try` block.

File: Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs [hunk 6]
```diff
@@ -30,93 +30,146 @@ internal class GatewayStoreModel : IStoreModelExtension, IDisposable
...
+                GatewayStoreModel.IsOperationSupportedByThinClient(request) &&
+                thinClientEndpoint != null;
+
+            DocumentServiceResponse response;
+
+            try
+            {
+                if (canUseThinClient)
+                {
+                    Uri physicalAddress = ThinClientStoreClient.IsFeedRequest(request.OperationType)
```
Comment (kundadebdatta) [inline Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs]: Resolve `physicalAddress ` just once for better code modularity. We can remove the duplicate on line `148`.

File: Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs (hunk 5 above)
Comment (aavasthy) [reply Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs]: then if thinClientEndpoint is null it will throw exception, with current implementation it just uses gateway mode

File: Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs (hunk 2 above)
Comment (aavasthy) [reply Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs]: yes can update it and fix the associated tests accordingly

File: Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs (hunk 6 above)
Comment (aavasthy) [reply Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs]: updated

File: Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs (hunk 4 above)
Comment (aavasthy) [reply Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs]: Added try-catch around this code like it was before

File: Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs (hunk 3 above)
Comment (aavasthy) [reply Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs]: Removed the check

File: Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs [hunk 7]
```diff
@@ -30,93 +30,146 @@ internal class GatewayStoreModel : IStoreModelExtension, IDisposable
...
+
+                    response = await this.thinClientStoreClient.InvokeAsync(
+                        request,
+                        request.ResourceType,
+                        physicalAddress,
+                        thinClientEndpoint,
+                        account.Id,
+                        this.clientCollectionCache,
+                        cancellationToken);
+                }
```
[the remark this answers is not in the document]
Comment (aavasthy) [reply Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs]: updated

File: Microsoft.Azure.Cosmos/src/DocumentClient.cs (hunk 1 above)
Comment (aavasthy) [reply Microsoft.Azure.Cosmos/src/DocumentClient.cs]: cleaned up

File: Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs (hunk 5 above)
Comment (kundadebdatta) [reply Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs]: I am wondering in which scenario would it be `null`. Are we not enabling thin client only when the gateway account reader has the thin client endpoints ?

File: Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs [hunk 8]
```diff
@@ -30,93 +30,141 @@ internal class GatewayStoreModel : IStoreModelExtension, IDisposable
 
         internal readonly GlobalEndpointManager endpointManager;
         private readonly DocumentClientEventSource eventSource;
-        internal readonly ConsistencyLevel defaultConsistencyLevel;
+        internal readonly ConsistencyLevel defaultConsistencyLevel;
+
+        private readonly bool enableThinClientMode;
```
Comment (kundadebdatta) [inline Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs]: Is `enableThinClientMode` and `userAgentContainer` needed in `GatewayStoreModel` ? If not - let's get them removed,

File: Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs [hunk 9]
```diff
@@ -30,93 +30,141 @@ internal class GatewayStoreModel : IStoreModelExtension, IDisposable
...
                 isPartitionLevelFailoverEnabled);
-
-            this.globalPartitionEndpointManager.SetBackgroundConnectionPeriodicRefreshTask(
+
+            this.enableThinClientMode = enableThinClientMode;
+            this.userAgentContainer = userAgentContainer;
+
+            if (enableThinClientMode)
+            {
+                this.thinClientStoreClient = new ThinClientStoreClient(
```
Comment (kundadebdatta) [inline Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs:70]: `userAgentContainer ` can be null here. Would it cause any issue in the `ThinClientStoreClient` processing ? If this is required, let's throw an `ArgumentException` with a proper message from the `GatewayStoreModel` constructor.

File: Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs (hunk 3 above)
Comment (kundadebdatta) [reply Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs]: A `null` `userAgentContainer` in the `ThincClientStoreClient` could cause unexpected behavior. If this is required, let's throw an `ArgumentException` with a proper message from the `GatewayStoreModel` constructor.

File: Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs [hunk 10]
```diff
@@ -492,27 +545,63 @@ internal static bool IsStoredProcedureCrudOperation(
...
+                    try
+                    {
+                        this.thinClientStoreClient.Dispose();
+                    }
+                    catch (Exception exception)
+                    {
+                        DefaultTrace.TraceWarning("Exception {0} thrown during dispose of HttpClient, this could happen if there are inflight request during the dispose of client",
+                            exception.Message);
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (kundadebdatta) [reply Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs]: This is a good suggestion @aavasthy.

File: Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs (hunk 8 above)
Comment (aavasthy) [reply Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs]: Yes they both are needed. The isThinClientEnabled/enableThinClientMode  is needed to check whether thinclient is to be used or gateway mode. userAgentContainer  is parameter needed for setting thinClientStoreClient.

File: Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (aavasthy) [reply Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs]: updated

File: Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs [hunk 11]
```diff
@@ -30,93 +30,141 @@ internal class GatewayStoreModel : IStoreModelExtension, IDisposable
 
         internal readonly GlobalEndpointManager endpointManager;
         private readonly DocumentClientEventSource eventSource;
-        internal readonly ConsistencyLevel defaultConsistencyLevel;
+        internal readonly ConsistencyLevel defaultConsistencyLevel;
+
+        private readonly bool isThinClientEnabled;
+        private readonly UserAgentContainer userAgentContainer;
```
Comment (kundadebdatta) [inline Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs]: I don't see any usage of the above members:

```
        private readonly bool isThinClientEnabled;
        private readonly UserAgentContainer userAgentContainer;
```

If they are not used in the `GatewayStpreModel`, let's get them removed.

File: Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs [hunk 12]
```diff
@@ -30,93 +30,141 @@ internal class GatewayStoreModel : IStoreModelExtension, IDisposable
...
+                serializerSettings,
                 isPartitionLevelFailoverEnabled);
-
-            this.globalPartitionEndpointManager.SetBackgroundConnectionPeriodicRefreshTask(
+
+            this.isThinClientEnabled = isThinClientEnabled;
+
+            if (isThinClientEnabled)
+            {
+                this.userAgentContainer = userAgentContainer ?? throw new ArgumentNullException(nameof(userAgentContainer));
```
Comment (kundadebdatta) [inline Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs]: Add a proper message in the Argument Exception.

File: Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs [hunk 13]
```diff
@@ -30,93 +30,141 @@ internal class GatewayStoreModel : IStoreModelExtension, IDisposable
...
+                    GatewayStoreModel.IsOperationSupportedByThinClient(request);
+
+                Uri physicalAddress = ThinClientStoreClient.IsFeedRequest(request.OperationType)
+                        ? this.GetFeedUri(request)
+                        : this.GetEntityUri(request);
+
+                if (canUseThinClient)
+                {
+                    Uri thinClientEndpoint = this.endpointManager.ResolveThinClientEndpoint(request);
+
```
Comment (kundadebdatta) [inline Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs:128]: Fix the spacings.

File: Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs (hunk 8 above)
Comment (kundadebdatta) [reply Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs]: Discussed offline. Let's get the private assignment from the `GatewayStoreModel`.

File: Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs (hunk 11 above)
Comment (aavasthy) [reply Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs]: updated

File: Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs (hunk 8 above)
Comment (aavasthy) [reply Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs]: updated

File: Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs (hunk 3 above)
Comment (aavasthy) [reply Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs]: updated

File: Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs (hunk 12 above)
Comment (aavasthy) [reply Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs]: updated

File: Microsoft.Azure.Cosmos/src/ThinClientStoreClient.cs [hunk 14]
```diff
@@ -40,7 +40,7 @@ public ThinClientStoreClient(
         {
             this.bufferProviderWrapperPool = new ObjectPool<BufferProviderWrapper>(() => new BufferProviderWrapper());
             this.isPartitionLevelFailoverEnabled = isPartitionLevelFailoverEnabled;
-            this.userAgentContainer = userAgentContainer;
+            this.userAgentContainer = userAgentContainer ?? throw new ArgumentNullException(nameof(userAgentContainer));
```
Comment (kundadebdatta) [inline Microsoft.Azure.Cosmos/src/ThinClientStoreClient.cs]: Can you add a test to assert on the null check, and validate the SDK initialization fails ?

File: Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs [hunk 15]
```diff
@@ -30,93 +30,136 @@ internal class GatewayStoreModel : IStoreModelExtension, IDisposable
...
+                httpClient,
+                this.eventSource,
+                serializerSettings,
                 isPartitionLevelFailoverEnabled);
-
-            this.globalPartitionEndpointManager.SetBackgroundConnectionPeriodicRefreshTask(
+
+            if (isThinClientEnabled)
+            {
+                this.thinClientStoreClient = new ThinClientStoreClient(
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs:70]: `ThinClientStoreClient' and 'GatewayStoreClient': Is relationship needs to be inheritence or composition?

File: Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs [hunk 16]
```diff
@@ -492,27 +535,65 @@ internal static bool IsStoredProcedureCrudOperation(
...
+                    catch (Exception exception)
+                    {
+                        DefaultTrace.TraceWarning("Exception {0} thrown during dispose of HttpClient, this could happen if there are inflight request during the dispose of client",
+                            exception.Message);
+                    }
+
+                    this.gatewayStoreClient = null;
+                }
+
+                if (this.thinClientStoreClient != null)
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs]: Better to move Thinclient one above/front before GWClient will incase of any future dependencies they might take,.

File: Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs (hunk 15 above)
Comment (aavasthy) [reply Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs:70]: Inheritance as we are overriding InvokeAsync method in child class

File: Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs (hunk 16 above)
Comment (aavasthy) [reply Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs]: updated

File: Microsoft.Azure.Cosmos/src/ThinClientStoreClient.cs (hunk 14 above)
Comment (aavasthy) [reply Microsoft.Azure.Cosmos/src/ThinClientStoreClient.cs]: Added test for checking useragent null eception.

File: Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs (hunk 5 above)
Comment (aavasthy) [reply Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs]: "Are we not enabling thin client only when the gateway account reader has the thin client endpoints " we are doing this part in DocumentClient for checking whether to use thinclient or gateway mode. 
However, this part gets it value based on   location cache resolution for this 
Uri thinClientEndpoint = this.endpointManager.ResolveThinClientEndpoint(request);

**Review by kirankumarkolli — APPROVED** [review]

**Review by kundadebdatta — APPROVED** [review]
